proglang/run-time-information-gathering
# :alembic: Produce message implementing the command pattern

This MR shows a very simple example of how the messages containing the incremental updates could look like.

Inspired by the [Command pattern](https://en.wikipedia.org/wiki/Command_pattern) and by the replication strategy of multiple DBs, for example https://redis.io/topics/replication.

The main idea is that the producer publishes the incremental updates in forms of _commands_ so that the consumer is able to rebuild the runTimeInfo object from these commands. The commands send as messages should contain all parameters needed to execute it.

### Example
The producer adds a new function container to the `runTimeInfo` object. It somehow executes the method `addFunctionContainer(functionId, functionContainer)`. Analogously, it publishes the message representing the command `add-function-container` with params `functionId` and `functionContainer`.

### Kafka implementation
The Kafka related logic is extracted in `utils/kafka.js`. It exports a single function `produceMessage(message)`. Internal properties such as `id` and `timestamp` are added to the message.

### How to test it
1) Start Kafka locally and start the consumer following the instructions in https://github.com/proglang/kafkajs-getting-started.

2) Set up a `.env` file with the correct values. Use `.env.example` as a template.

2) Run the analysis. Due to internal Jalangi logics, `KafkaJS` is being instrumented, which throws an error. In order to disable the instrumentation the blacklisting feature has to be used:

```bash
$ npm run --silent generate -- tests/calculator/calculator.js blacklisted.json
```

3) Check the consumer you started on point 1. It will start printing the produced messages. They will look similar to this:
```
Receiving message:
{"command":"add-function-container","data":{"functionId":"functionId_3","functionContainer":"{\"functionId\":\"functionId_3\",\"functionName\":\"sum\",\"isConstructor\":false,\"args\":{},\"declarationEnclosingFunctionId\":-1,\"returnTypeOfs\":[],\"functionIid\":49,\"requiredModule\":\"\",\"isExported\":false,\"constructedBy\":\"functionId_1\"}"},"id":"mQQiMB8H1BKbq9vBtrtUd","timestamp":"2021-12-05T21:20:05.952Z"}
```
